### ISSUE TITLE
File descriptor leak with HTTPS proxies via CONNECT

### ISSUE THREAD

**Issue by ianpreston** [shazow/urllib3#548]
It looks like when making HTTPS requests through an HTTP proxy, there is a file descriptor leak if an error occurs (e.g. socket timeout or 404 error) during httplib.HTTPConnection._tunnel. Tested in versions 1.9, 1.9.1, 1.10.

The way I've been testing is to use netcat to listen for connections but not respond to CONNECT, eg.:

    $ nc -l localhost 8000

Then making a request with that "proxy":

    In [1]: import urllib3
    In [2]: pm = urllib3.ProxyManager('http://localhost:8000/')
    In [3]: r = pm.request('GET', 'https://httpbin.org/ip', timeout=1.0)
    ---------------------------------------------------------------------------
    MaxRetryError                             Traceback (most recent call last)

At that point, `lsof` shows multiple connections (one per retry) in CLOSE_WAIT state, which stay that way permanently.

It looks like the issue is related to a circular reference preventing a socket from being cleaned up properly. For example, this commit fixes the issue: https://github.com/ianpreston/urllib3/commit/a0953a93c877468d9b60a6785c767c5ac31d0dc2 , as does wrapping the `raise` statements in `Retry.increment` with a "finally: del _stacktrace".

The workaround above is working for now but I'd like to find the root cause, or at least a workaround that doesn't break stack traces. Any idea where to start looking? Thanks

**Comment by shazow** [comment]
@ianpreston Great find, thank you!

Honestly I'm not very experienced with finding memory leaks in Python. @sigmavirus24 @Lukasa?

Wrapping with `finally: del _stacktrace` sounds reasonable. Another approach is we could use weakrefs where appropriate. Would you like to send a PR to this effect? :)

**Comment by Lukasa** [comment]
I'm going to admit to a degree of confusion here.

How can the stacktrace object be keeping the socket alive? That stacktrace object passed to the retry increment method doesn't get saved on the object, it only gets thrown. Unless `six.reraise` is holding on to it somewhere, I don't see the circular reference. Can you enlighten me?

**Comment by ianpreston** [comment]
@Lukasa My understanding is that each frame of the traceback object returned by `sys.exc_info` holds references to the locals of that stack frame. So if it gets assigned to a local variable, there will be a `locals -> frame -> locals` loop.

My thinking is that there's a socket assigned to a local variable somewhere (possibly httplib) that's being opened but not properly closed. Since CPython closes file descriptors when their refcount hits zero, removing the cycle causes the sockets to be closed. This is all speculation, though. I could be totally wrong on that.

I'll have some time later this week to dig a little deeper. Will at least submit a PR with `finally: del _stacktrace` wrappers.

**Comment by sigmavirus24** [comment]
Can someone see if `objgraph` agrees with @ianpreston's understanding above?

**Comment by ianpreston** [comment]
Ok, think I found it.


The [httplib.HTTPConnection._tunnel](https://hg.python.org/cpython/file/687b970c8ba9/Lib/httplib.py#l776) method creates an HTTPResponse instance, which [calls makefile](https://hg.python.org/cpython/file/687b970c8ba9/Lib/httplib.py#l350) on the HTTPConnection's socket. Though urllib3 will call `close` on the HTTPConnection on error, nothing will actually close the HTTPResponse's file-like object.

Since the HTTPResponse is a local variable, it would normally be expected to be deleted once `_tunnel` pops off the stack, but the stacktrace object holds a reference to `_tunnel`'s locals, so the refcount won't drop to zero.

@sigmavirus24 Sorry if my explaination was unclear. I wrote a gist which hopefully should illustrate what I was trying to say more clearly: https://gist.github.com/ianpreston/12cbf93196eddc6dd5cd


I submitted a pull request with a fix: https://github.com/shazow/urllib3/pull/549 . The tests pass, at least on my machine. Let me know if I did it right, this is my first pull req to urllib3.
